lib/codegen/DecisionParser.cpp

```diff
@@ -0,0 +1,161 @@
+#include "kllvm/codegen/DecisionParser.h"
+
+#include <yaml-cpp/yaml.h>
+
+#include <stack>
+
+namespace kllvm {
+
+class DTPreprocessor {
+private:
+  std::vector<std::string> constructors;
+  std::map<std::vector<int>, std::string> occurrences;
+  const llvm::StringMap<KOREObjectSymbol *> &syms;
+  int counter;
+  int functionDepth;
+  KOREObjectSymbol *dv;
+
+  enum Kind {
+    Switch, SwitchLit, Function, Leaf, Fail, Swap
+  };
+
+  static Kind getKind(YAML::Node node) {
+    if (node.IsScalar()) return Fail;
+    if (node["bitwidth"]) return SwitchLit;
+    if (node["specializations"]) return Switch;
+    if (node["action"]) return Leaf;
+    if (node["swap"]) return Swap;
+    return Function;
+  }
+
+public:
+  DTPreprocessor(int numSubjects, const llvm::StringMap<KOREObjectSymbol *> &syms) 
+      : syms(syms), counter(0), functionDepth(0) {
+    for (int i = numSubjects - 1; i >= 0; --i) {
+      std::string name = "subject" + std::to_string(i);
+      constructors.push_back(name);
+      occurrences[{i+1}] = name;
+    }
+    dv = KOREObjectSymbol::Create("\\dv");
+  }
+
+  DecisionNode *swap(YAML::Node node) {
+    int idx = node["swap"][0].as<int>();
+    std::string tmp = constructors[constructors.size() - 1 - idx];
+    constructors[constructors.size() - 1 - idx] = constructors[constructors.size() - 1];
+    constructors[constructors.size() - 1] = tmp;
+    auto result = (*this)(node["swap"][1]);
+    return result;
+  }
+
+  DecisionNode *function(YAML::Node node) {
+    std::string function = node["function"].as<std::string>();
+    std::string hookName = node["sort"].as<std::string>();
+    SortCategory cat = KOREObjectCompositeSort::getCategory(hookName);
+
+    std::string binding = "_" + std::to_string(counter++);
+    occurrences[{functionDepth++, 0}] = binding;
+    constructors.push_back(binding);
+
+    auto child = (*this)(node["next"]); 
+
+    functionDepth--;
+
+    auto result = FunctionNode::Create(binding, function, child, cat);
+    
+    YAML::Node vars = node["args"];
+    for (auto iter = vars.begin(); iter != vars.end(); ++iter) {
+      auto var = *iter;
+      auto occurrence = var.as<std::vector<int>>();
+      result->addBinding(occurrences[occurrence]);
+    }
+    return result;
+  }
+
+  DecisionNode *switchCase(Kind kind, YAML::Node node) {
+    YAML::Node list = node["specializations"];
+    std::string name = constructors.back();
+    constructors.pop_back();
+    auto result = SwitchNode::Create(name);
+    auto occurrence = node["occurrence"].as<std::vector<int>>();
+    for (auto iter = list.begin(); iter != list.end(); ++iter) {
+      auto _case = *iter;
+      std::vector<std::string> copy = constructors;
```

Comment:
It looks strange to set up state for the child by modifying the field constructors, and then reset it from a copy in the local variable, rather than setting up the list for the child in the local variable.